### PR TITLE
Generate UpdateStatus for workflow and CronWorkflow and install CronW…

### PR DESCRIPTION
**Pull request by sdminonne**
Generate UpdateStatus for workflow and CronWorkflow. 
Install CronWorkflow CRD
Create skeleton for cronWorkflowController

### CODE REVIEW

**Comment by clamoriniere1A** [comment]
@sdminonne you need to add the CRD creation into the Helm chart.

File: app/config.go [hunk 1]
```diff
@@ -34,5 +36,7 @@ func NewWorkflowControllerConfig() *Config {
 // AddFlags add cobra flags to populate Config
 func (c *Config) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&c.KubeConfigFile, "kubeconfig", c.KubeConfigFile, "Location of kubecfg file for access to kubernetes master service")
-	fs.StringVar(&c.ListenHTTPAddr, "addr", "0.0.0.0:8086", "listen address of the http server which serves kubernetes probes and prometheus endpoints")
+	fs.StringVar(&c.ListenHTTPAddr, "addr", "0.0.0.0:8086", "Listen address of the http server which serves kubernetes probes and prometheus endpoints")
+	fs.BoolVar(&c.EnableCronWorkflow, "enable-cron-workflow", false, "Enable CronWorkflow controller")
```
Comment (clamoriniere1A) [inline app/config.go]: Why do you want to propose the possibility to activate or not CronWorkflow?

File: pkg/api/cronworkflow/v1/register.go [hunk 2]
```diff
@@ -1,11 +1,10 @@
 package v1
 
 import (
+	"github.com/amadeusitgroup/workflow-controller/pkg/api/workflow"
```
Comment (clamoriniere1A) [inline pkg/api/cronworkflow/v1/register.go]: please reorganize imports.

File: pkg/api/cronworkflow/v1/types.go [hunk 3]
```diff
@@ -102,7 +100,7 @@ type CronWorkflowStatus struct {
 type WorkflowTemplateSpec struct {
 	// Standard object's metadata of the workflows created from this template.
 	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
-	// +optional
+	// +optionalfr
```
Comment (clamoriniere1A) [inline pkg/api/cronworkflow/v1/types.go]: is this change a typo? what does mean `+optionalfr` ?

File: pkg/controller/cronworkflow.go [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+		}
+		return
+	}
+	w.enqueue(workflow)
+}
+
+func (w *CronWorkflowController) onUpdateCronWorkflow(oldObj, newObj interface{}) {
+	workflow, ok := newObj.(*cwapi.CronWorkflow)
+	if !ok {
+		glog.Errorf("Expected workflow object. Got: %+v", newObj)
```
Comment (clamoriniere1A) [inline pkg/controller/cronworkflow.go]: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: pkg/controller/cronworkflow.go [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+		return
+	}
+	w.enqueue(workflow)
+}
+
+func (w *CronWorkflowController) onDeleteCronWorkflow(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Unable to get key for %#v: %v", obj, err)
```
Comment (clamoriniere1A) [inline pkg/controller/cronworkflow.go]: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: pkg/controller/workflow_test.go [hunk 6]
```diff
@@ -15,7 +15,9 @@ import (
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/cache"
 
+	"github.com/amadeusitgroup/workflow-controller/pkg/api/workflow"
 	wapi "github.com/amadeusitgroup/workflow-controller/pkg/api/workflow/v1"
+
```
Comment (clamoriniere1A) [inline pkg/controller/workflow_test.go]: this line can be removed

File: test/e2e/workflows.go [hunk 7]
```diff
@@ -10,7 +10,9 @@ import (
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
 
+	"github.com/amadeusitgroup/workflow-controller/pkg/api/workflow"
 	wapi "github.com/amadeusitgroup/workflow-controller/pkg/api/workflow/v1"
+
```
Comment (clamoriniere1A) [inline test/e2e/workflows.go]: this line can be removed

File: pkg/api/cronworkflow/v1/types.go (hunk 3 above)
Comment (sdminonne) [reply pkg/api/cronworkflow/v1/types.go]: cruft :)

File: app/config.go (hunk 1 above)
Comment (sdminonne) [reply app/config.go]: hmmm... At the beginning I thought it was a good idea. I think you're right. I'm going to enable by default.

File: pkg/api/cronworkflow/v1/register.go (hunk 2 above)
Comment (sdminonne) [reply pkg/api/cronworkflow/v1/register.go]: done thanks

**Comment by sdminonne** [comment]
@clamoriniere1A PTA

File: app/workflow-controller.go [hunk 8]
```diff
@@ -61,58 +64,78 @@ func initKubeConfig(c *Config) (*rest.Config, error) {
...
 		go handleSignal(cancelFunc)
 		c.kubeInformerFactory.Start(ctx.Done())
 		c.workflowInformerFactory.Start(ctx.Done())
 		c.runGC(ctx)
 		go c.runHTTPServer(ctx)
-		c.controller.Run(ctx)
+		c.cronWorkflowInfomerFactory.Start(ctx.Done())
+		c.workflowInformerFactory.Start(ctx.Done())
+		c.workflowController.Run(ctx)
+		// c.cronWorkflowController.Run(ctx): TODO: activates this
```
Comment (clamoriniere1A) [inline app/workflow-controller.go]: can be uncomment

File: pkg/client/client.go [hunk 9]
```diff
@@ -15,27 +16,28 @@ import (
 
 	"github.com/golang/glog"
 
+	cronworkflowV1 "github.com/amadeusitgroup/workflow-controller/pkg/api/cronworkflow/v1"
 	"github.com/amadeusitgroup/workflow-controller/pkg/api/workflow"
-	v1 "github.com/amadeusitgroup/workflow-controller/pkg/api/workflow/v1"
+	workflowV1 "github.com/amadeusitgroup/workflow-controller/pkg/api/workflow/v1"
```
Comment (clamoriniere1A) [inline pkg/client/client.go]: empty line can be removed

File: pkg/controller/cronworkflow.go [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	cwapi "github.com/amadeusitgroup/workflow-controller/pkg/api/cronworkflow/v1"
+	wapi "github.com/amadeusitgroup/workflow-controller/pkg/api/workflow/v1"
+
```
Comment (clamoriniere1A) [inline pkg/controller/cronworkflow.go]: empty line can be removed

File: pkg/controller/cronworkflow.go [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+	wclientset "github.com/amadeusitgroup/workflow-controller/pkg/client/clientset/versioned"
+	cwlisters "github.com/amadeusitgroup/workflow-controller/pkg/client/listers/cronworkflow/v1"
+	wlisters "github.com/amadeusitgroup/workflow-controller/pkg/client/listers/workflow/v1"
+)
+
+// CronWorkflowControllerConfig contains info to customize Workflow controller behaviour
+type CronWorkflowControllerConfig struct {
+	RemoveIvalidCronWorkflow bool
+	NumberOfThreads          int
```
Comment (clamoriniere1A) [inline pkg/controller/cronworkflow.go]: it's not really a number of threads but the number of workers

File: pkg/controller/cronworkflow.go [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+
+	updateHandler func(*cwapi.CronWorkflow) error // callback to upate Workflow. Added as member for testing
+	syncHandler   func(string) error
+
+	Recorder record.EventRecorder
+	config   CronWorkflowControllerConfig
+}
+
+// NewCronWorkflowController creates and initializes the CronWorkflowController instance
+func NewCronWorkflowController(
```
Comment (clamoriniere1A) [inline pkg/controller/cronworkflow.go]: I think this function should take a CronWorkflowControllerConfig as a parameter

File: app/workflow-controller.go (hunk 8 above)
Comment (sdminonne) [reply app/workflow-controller.go]: No not yet

File: pkg/client/client.go (hunk 9 above)
Comment (sdminonne) [reply pkg/client/client.go]: right

File: pkg/controller/cronworkflow.go (hunk 12 above)
Comment (sdminonne) [reply pkg/controller/cronworkflow.go]: right, but WorkflowControlleris the same. Going to do in another PR for both

**Comment by sdminonne** [comment]
@clamoriniere1A  comments addressed. Would like to squash and merge. Mind quickly relook at it?

**Pull request merged**